Grid selection refactoring

thesiv opened the pull request.

thesiv committed:
Remove unused wxGrid member variable

thesiv committed:
Make wxGridSelection to be non-friend to wxGrid

Improve integrity of the wxGridSelection internal data by removing `friend`
declaration.

thesiv committed:
Clear the selection in wxGrid::SetTable

The selection in m_selection was not considered at all so it was worked not
correctly any way.

thesiv committed:
Implement wxGridBlockCoords class

wxGridBlockCoords represents a location of a block of cells in the grid.

thesiv committed:
Refactor wxGridSelection to store selection as blocks only

Store all types of selection with an array of blocks instead of arrays of
cells, blocks, rows and columns.

It (hopefully) simplifies the code and allows us to implement editing of
the last selection block much easier.

thesiv committed:
Fix SelectEmptyGrid wxGrid unit test

Use CATCH nested sections correctly.

thesiv committed:
Edit the current wxGrid selection block

Really edit the current selection block instead of storing the temporary
information about the current selection and applying it on releasing Shift
key or LKM.

thesiv committed:
Make wxGrid column selecting more user friendly

thesiv committed:
Make wxGrid row selecting more user friendly

thesiv committed:
Make wxGrid cells selecting by mouse more user friendly

thesiv committed:
Refactor wxGrid::MakeCellVisible() to scroll in only one direction

Allow -1 for a row or a column single parameter.

thesiv committed:
Improve rows, columns and cells deselection in wxGrid

Use DeselectBlock() instead of ToggleCellSelection() to improve execution
speed and make the code more clean.

thesiv committed:
Implement wxGrid selection blocks iterating interface

thesiv committed:
Remove wxGridSelection::ToggleCellSelection()

Remove the function because it's not usefull anymore and used only in one
place.

thesiv committed:
Do not merge wxGrid selection blocks

Remove blocks merging because it complicates the behavior at the user point
of view and the code.

thesiv committed:
Move the wxGrid cursor to the corner of the last selection

Move the wxGrid cursor to the corner of the last selection when splitting
the selection block on click with Ctrl.

vadz commented on include/wx/generic/private/grid.h:
```diff
@@ -793,6 +793,19 @@ class wxGridDirectionOperations
+    }
+
+    // Make the coordinates with the other component value of -1.
+    wxGridCellCoords MakeLineCoords(const wxGridCellCoords& coords) const
```
I think this name might be more clear:
```suggestion
    wxGridCellCoords MakeWholeLineCoords(const wxGridCellCoords& coords) const
```

vadz commented on interface/wx/grid.h:
```diff
@@ -1816,6 +1816,167 @@ class wxGridCellCoords
+
+    An object of this class is a coordinates of the left top and the bottom right
+    corners of a block.
+ */
```
Should have `@since 3.1.4`

vadz commented on interface/wx/grid.h:
```diff
@@ -1816,6 +1816,167 @@ class wxGridCellCoords
+/**
+    Represents coordinates of a block of cells in the grid.
+
+    An object of this class is a coordinates of the left top and the bottom right
```
```suggestion
    An object of this class contains coordinates of the left top and the bottom right
```

vadz commented on interface/wx/grid.h:
```diff
@@ -1816,6 +1816,167 @@ class wxGridCellCoords
+    An object of this class is a coordinates of the left top and the bottom right
+    corners of a block.
+ */
+class WXDLLIMPEXP_CORE wxGridBlockCoords
```
```suggestion
class wxGridBlockCoords
```

vadz commented on interface/wx/grid.h:
```diff
@@ -1816,6 +1816,167 @@ class wxGridCellCoords
+        Default constructor initializes the object to invalid state.
+
+        Initially the coordinates are invalid (-1) and so operator!() for an
+        uninitialized wxGridBlockCoords returns false.
```
```suggestion
        uninitialized wxGridBlockCoords returns @true.
```

vadz commented on interface/wx/grid.h:
```diff
@@ -1816,6 +1816,167 @@ class wxGridCellCoords
+    /**
+        Checks whether the cells block is invalid.
+
+        Returns false only if all components are -1. Notice that if one
```
```suggestion
        Returns @true only if all components are -1. Notice that if one
```

vadz commented on interface/wx/grid.h:
```diff
@@ -1816,6 +1816,167 @@ class wxGridCellCoords
+        Checks whether the cells block is invalid.
+
+        Returns false only if all components are -1. Notice that if one
+        of components (but not all) is -1, this method returns true even if
```
```suggestion
        of the components (but not all) is -1, this method returns @false even if
```

vadz commented on interface/wx/grid.h:
```diff
@@ -1816,6 +1816,167 @@ class wxGridCellCoords
+
+/**
+    The iterator for grid selection blocks.
+ */
```
Also misses `@since`.

vadz commented on interface/wx/grid.h:
```diff
@@ -1816,6 +1816,167 @@ class wxGridCellCoords
+    wxGridBlockCoords *m_end;
+    wxGridBlockCoords *m_it;
+
+    friend class wxGrid;
```
Private parts don't need to/shouldn't be present in the interface files unless it's particularly useful for some reason (e.g. above it could be considered to be useful because it shows that `wxGridBlockCoords` is just a trivial POD type).
```suggestion
```

vadz commented on interface/wx/grid.h:
```diff
@@ -4409,6 +4577,9 @@ class wxGrid : public wxScrolledCanvas
         array and GetSelectedCols() returns an array containing one element).
+
+        @since 3.1.4 The function can be slow for the big grids, use
+        GetSelectionIterator in the new code.
```
I think it needs parentheses to be auto-linked in the generated docs.
```suggestion
        GetSelectionIterator() in the new code.
```

vadz commented on interface/wx/grid.h:
```diff
@@ -4409,6 +4577,9 @@ class wxGrid : public wxScrolledCanvas
         a million of entries in this function, instead it returns an empty
         array and GetSelectedCols() returns an array containing one element).
+
+        @since 3.1.4 The function can be slow for the big grids, use
```
Ha, here is a rare example of a superfluous `@since`. It really shouldn't be used for the functions that existed before 3.1.4. If the function wasn't slow before, you need to write in full words that it became less efficient in 3.1.4 and so the new function should be used instead. But I think it hasn't actually changed, so this could be just removed:
```suggestion
        The function can be slow for the big grids, use
```

vadz commented on interface/wx/grid.h:
```diff
@@ -4420,6 +4591,9 @@ class wxGrid : public wxScrolledCanvas
         or being selected in virtue of all of their cells being selected
         individually, please see GetSelectedCells() for more details.
+
+        @since 3.1.4 The function can be slow for the big grids, use
```
Same as above (and also 3 more occurrences of the same below).

vadz reviewed (commented):
I haven't checked the changes to the code itself yet (will try to do soon), but I'm already publishing the comments so far to let you know about them.

vadz commented on src/generic/gridsel.cpp:
```diff
@@ -398,517 +288,262 @@ void wxGridSelection::SelectBlock( int topRow, int leftCol,
+    default:
+        wxFAIL_MSG("unknown selection mode");
+        wxFALLTHROUGH;
+
```
It would be best to avoid having `default` here, as this inhibits warnings about not handling some enum elements. The best would be to have something like this:
```
int allowed = -1;
switch (m_selectionMode) {
    case SelectCells | SelectRows | SelectColumns: ...; allowed = 1; break;
    case SelectRowsOrColumns: allowed = 0; break;
}

wxASSERT_MSG(allowed != -1, "unknown selection mode?" );
if ( !allowed )
     return;
```

vadz commented on src/generic/gridsel.cpp:
```diff
@@ -398,517 +288,262 @@ void wxGridSelection::SelectBlock( int topRow, int leftCol,
+
+    case wxGrid::wxGridSelectRows:
+        leftCol = 0;
+        rightCol = m_grid->GetNumberCols() - 1;
```
Are we sure the grid has columns here (and rows below)?

vadz replied:
OK, it seems this doesn't matter because `Select()` checks for this anyhow.

vadz commented on src/generic/gridsel.cpp:
```diff
@@ -398,517 +288,262 @@ void wxGridSelection::SelectBlock( int topRow, int leftCol,
+
+    count = m_selection.size();
+    for ( n = 0; n < count; n++ )
+    {
```
If the body of this loop could be refactored in another `wxGridBlockCoords` method or a private function (i.e. a function in a private header) that could be unit tested directly, it would help a lot as it's really difficult to be sure that there are no bugs here right now.

vadz commented on src/generic/gridsel.cpp:
```diff
@@ -398,517 +288,262 @@ void wxGridSelection::SelectBlock( int topRow, int leftCol,
     }
 }
 
+
```
Very minor, but this shouldn't be added.

vadz commented on src/generic/gridsel.cpp:
```diff
@@ -398,517 +288,262 @@ void wxGridSelection::SelectBlock( int topRow, int leftCol,
-        coords2 = m_blockSelectionBottomRight[n];
-        m_blockSelectionTopLeft.RemoveAt(n);
-        m_blockSelectionBottomRight.RemoveAt(n);
+        wxGridBlockCoords& block = m_selection[n];
```
```suggestion
        const wxGridBlockCoords& block = m_selection[n];
```

vadz commented on src/generic/gridsel.cpp:
```diff
@@ -975,222 +572,330 @@ void wxGridSelection::ClearSelection()
-    if ( !m_grid->GetNumberCols() )
-        m_rowSelection.Clear();
+    wxArrayInt result;
+    result.Alloc(uniqueCols.size());
```
```suggestion
    result.reserve(uniqueCols.size());
```
Prefer using standard-like functions when possible.

vadz commented on src/generic/gridsel.cpp:
```diff
@@ -975,222 +572,330 @@ void wxGridSelection::ClearSelection()
+    wxIntSet::iterator it;
+    for( it = uniqueCols.begin(); it != uniqueCols.end(); ++it )
+    {
+        result.Add(*it);
```
```suggestion
        result.push_back(*it);
```

vadz commented on src/generic/gridsel.cpp:
```diff
@@ -975,222 +572,330 @@ void wxGridSelection::ClearSelection()
+    wxArrayInt result;
+    result.Alloc(uniqueCols.size());
+    wxIntSet::iterator it;
+    for( it = uniqueCols.begin(); it != uniqueCols.end(); ++it )
```
```suggestion
    for ( wxIntSet::iterator it = uniqueCols.begin(); it != uniqueCols.end(); ++it )
```

vadz commented on src/generic/gridsel.cpp:
```diff
@@ -975,222 +572,330 @@ void wxGridSelection::ClearSelection()
+    return coords;
+}
+
+wxArrayInt wxGridSelection::GetRowSelection() const
```
Same comments as for columns selection below.

vadz commented on src/generic/gridsel.cpp:
```diff
@@ -975,222 +572,330 @@ void wxGridSelection::ClearSelection()
+{
+    if ( m_selectionMode == wxGrid::wxGridSelectRows )
+        return wxArrayInt();
+
```
I think some comment describing what's going on here would be welcome, e.g.
```
// For compatibility with the existing code, try reconstructing the selected
// columns from the selected blocks we store internally. Of course, this only
// works well in the column selection mode in which the user can only select
// the entire columns in the first place, as otherwise it's difficult to
// efficiently determine that a column is selected because all of its cells
// were selected one by one. But this should work well enough in practice and
// is, anyhow, the best we can do.
```
Perhaps you could put this comment before both row/column selection functions to avoid repeating it.

vadz commented on src/generic/gridsel.cpp:
```diff
@@ -975,222 +572,330 @@ void wxGridSelection::ClearSelection()
+        return wxGridCellCoordsArray();
+
+    wxGridCellCoordsArray cells;
+    const size_t count = m_selection.size();
```
Should call `reserve()`.

vadz commented on src/generic/gridsel.cpp:
```diff
@@ -975,222 +572,330 @@ void wxGridSelection::ClearSelection()
+        if ( block.GetTopRow() == block.GetBottomRow() &&
+             block.GetLeftCol() == block.GetRightCol() )
+        {
+            cells.Add(block.GetTopLeft());
```
```suggestion
            cells.push_back(block.GetTopLeft());
```

vadz commented on src/generic/gridsel.cpp:
```diff
@@ -975,222 +572,330 @@ void wxGridSelection::ClearSelection()
+    return cells;
+}
+
+wxGridCellCoordsArray wxGridSelection::GetBlockSelectionTopLeft() const
```
Same comments as in the function above.

vadz commented on src/generic/gridsel.cpp:
```diff
@@ -975,222 +572,330 @@ void wxGridSelection::ClearSelection()
 
-    count = m_colSelection.GetCount();
-    for ( n = 0; n < count; n++ )
+bool wxGridSelection::EditCurrentBlock(int startRow, int startCol,
```
It's still a bit difficult to understand what does this function do, but I don't think "Edit" is a good name. Perhaps it should be "Extend"? Or "ExtendOrReplace"?

In fact, the fact that this function does something apparently complicated makes me wonder if we don't need 2 functions instead, one for extending the current block and another for starting a new one. Could this be simpler?

vadz reviewed (commented):
I'm done with reviewing this now and globally it looks good, but I'd really like to extract more complicated functions so that they could be tested.

And I know that there are a lot of comments, but please go through all of them, I won't be able to re-review the next version as thoroughly again, so please check that you don't miss/forget some of them.

TIA!

thesiv committed:
Remove wxEVT_GRID_RANGE_SELECT handling by wxGrid::SendEvent()

VZ: [...] this is not used by wxGrid itself and SendEvent() is not part
of the public API, so we never made any promises about being able to call
it with wxEVT_GRID_RANGE_SELECT.

thesiv committed:
Fix the test where wxGrid scrolling when selecting cells

thesiv commented:
Fixed a lot of things, but not everything.

vadz commented:
Force pushed after interactive rebase.

vadz committed:
Define wxGridSelectionRange::iterator in terms of vector iterator

This fixes compilation when using std::vector<> implementation for
wxVector (i.e. with wxUSE_STD_CONTAINERS==1).

vadz committed:
Make temporary variable in GetSelectionRange() const

We don't need to modify the selection blocks here.

No real changes.

vadz committed:
Simplify wxGridSelectionRange to provide only iterators

This class was a strange hybrid of a container/view/range and iterator,
as it both provided begin()/end() container-like methods and
iterator-like methods for dereferencing/advancing.

Simplify this by removing the latter part and making this class really
just a range, with its own iterator class in order to avoid leaking the
exact type of the iterator used in the API.

Note that while it's now completely trivial, it is still useful as it
isolates the application code from the vector used to store the selected
blocks currently and will allow to change internal representation in the
future without breaking the existing code.

vadz commented on include/wx/generic/grid.h:
```diff
@@ -742,10 +742,148 @@ class WXDLLIMPEXP_CORE wxGridCellCoords
+class wxGridSelectionIterator
+{
+public:
+    typedef wxGridBlockCoords* iterator;
```
I've defined a nested iterator class forwarding to `wxVector<wxGridBlockCoords>::const_iterator` to resolve this.

vadz commented on include/wx/generic/gridsel.h:
```diff
@@ -58,19 +62,60 @@ class WXDLLIMPEXP_CORE wxGridSelection
     void UpdateRows( size_t pos, int numRows );
     void UpdateCols( size_t pos, int numCols );
 
+    // Edit the current selection block or select a new one.
```
This was extensively discussed and the function was renamed to `ExtendOrCreateCurrentBlock()~ and comment updated to address this.

vadz committed:
Rename wxGridBlockCoords::ContainsCell() and move it inline

Make the function name more grammatically correct.

No real changes.

vadz committed:
Rename and simplify wxGridBlockCoords::ContainsBlock()

Change the return type of this function to a simple and clear bool
instead of 3-valued int requiring a special explanation. This is simpler
and not any less efficient as checking for whether one block contains
another or the other one contains this one are separate operations
anyhow.

Rename the function to a more grammatically correct name.

Also move it inline as it's now trivial enough for this to be worth it.

vadz committed:
Add a comment about wxGridSelection::IsInSelection() complexity

No real changes.

vadz commented on src/generic/gridsel.cpp:
```diff
@@ -46,66 +167,22 @@ wxGridSelection::wxGridSelection( wxGrid * grid,
-    // Now check whether the given cell is
-    // contained in one of the selected blocks.
-    count = m_blockSelectionTopLeft.GetCount();
+    // Check whether the given cell is contained in one of the selected blocks.
```
For now I've just added a comment, it doesn't seem worth it to do anything else/more.

vadz committed:
Remove outdated comment from wxGridSelection code

Don't mention variable which doesn't exist any more.

vadz committed:
Simplify wxGrid::SetSelectionMode() selection updating logic

Don't try to extend the existing selected blocks to rows/columns, this
contradicts the documented behaviour which is to discard the selected
blocks that become invalid in the new mode.

Do handle switching to wxGridSelectRowsOrColumns mode, as there doesn't
seem to be any reason not to.

Update the tests to check for the expected selection update behaviour.

vadz commented on src/generic/gridsel.cpp:
```diff
@@ -116,6 +193,8 @@ void wxGridSelection::SetSelectionMode( wxGrid::wxGridSelectionModes selmode )
     if (selmode == m_selectionMode)
         return;
 
+    // TODO: wxGridSelectRowsOrColumns?
```
Added support for it now.

vadz committed:
Increase the size of the windows in the grid sample

Make larger part of the grid visible initially and also show more lines
in the log window.

No real changes, this is purely cosmetic.

vadz committed:
Refactor selection expansion code to use actual wxKeyboardState

Switch from using just "bool expandSelection" in the grid functions
(possibly) extending the current selection to using the full
wxKeyboardState.

This allows to pass it to ExtendOrCreateCurrentBlock() and slightly
simplify the code by using DoMoveCursorFromKeyboard().

vadz commented on src/generic/grid.cpp:
```diff
@@ -7770,7 +7719,20 @@ wxGrid::DoMoveCursorByBlock(bool expandSelection,
     if ( expandSelection )
     {
-        UpdateBlockBeingSelected(m_currentCellCoords, coords);
+        // TODO: Select the next block every time (not the same as now).
```
I still have no idea about this one. I've refactored the code to fix the "TODO: pass keyboard state" part of this comment in the latest version, but what does the "same block" mean remains unclear.

Is this related to the currently wrong behaviour of `Shift-Ctrl-<Dir>`? Instead of extending the current block, it continues selection from the current cell, which is surprising and different from the standard spreadsheet behaviour.

vadz replied:
I've refactored the code to use the same initial cell coords in `DoMoveCursorByBlock()` as in `DoMoveCursor()` itself and this seems to fix the behaviour. Please let me know if you still something wrong from the user point of view here.

vadz committed:
Remove unused private BlockContain() function

This was probably left over after wxGridBlockCoords::ContainsBlock()
introduction.

vadz committed:
Some minor comments wording changes in wxGridSelection

Try to better explain the behaviour of the methods of this class.

vadz committed:
Use the same selection expansion logic for Shift-Ctrl-cursor

Expanding the selection from keyboard with Ctrl pressed should move in
the same way Ctrl-cursor does, but use the same selection anchor as
Shift-cursor does instead of always using the current cell.

This makes the expansion work much more intuitively in the grid, e.g.
pressing Shift-Ctrl-Down in

    1 2
    3 4

grid when 1 and 2 are selected now selects all the cells instead of
selecting 1 and 3 as it did before.

vadz committed:
Add MSVS debug visualizers for wxGrid{Cell,Block}Coords

This makes debugging wxGrid code much more pleasant.

vadz commented on src/generic/gridsel.cpp:
```diff
@@ -975,222 +359,331 @@ void wxGridSelection::ClearSelection()
+
+    if ( blockEnd.GetRow() != -1 )
+    {
+        if ( newBlock.GetTopRow() == blockStart.GetRow() )
+        {
+            newBlock.SetBottomRow(blockEnd.GetRow());
+            editBlock = true;
+        }
+        else if ( newBlock.GetBottomRow() == blockStart.GetRow() )
+        {
+            newBlock.SetTopRow(blockEnd.GetRow());
+            editBlock = true;
+        }
```
Just FYI, all this doesn't work at all for column selection mode (and, of course, the reversed version below doesn't work for the much more often used row selection mode), it results in having non-column selection blocks as it adjusts rows to be different from the 0/last.

vadz replied:
Moreover, the selection misbehaves, even in cell selection mode, when neither of the 2 conditions are satisfied. There really must be something in the (currently missing) `else` branch here.

vadz replied:
For the reference, this "something" is the code to expand the current block to the newly specified block in this direction. This is required to allow Shift-clicking on the header to extend the existing selection correctly and only expanding, but not shrinking, the current block is required for Shift-arrows to preserve the full line selection afterwards.

vadz committed:
Slightly clarify ExtendOrCreateCurrentBlock() code

Don't use long-lived non-const "block" reference, this makes it
difficult to see that it's never changed until the very end of the
function, when it's updated.

Compare "blockStart" with "block" itself instead of comparing it with
"newBlock" which is initially its copy but gets modified later (however
these modifications don't affect the comparisons involving it).

Add a comment explaining why are we doing what we do here.

vadz committed:
Show more detailed information about selection in the grid sample

Show the selected cells/rows/columns/blocks instead of just showing
their number, as we now have an efficient (i.e. taking time proportional
to the number of selected blocks and not to the number of the individual
cells, rows or columns) way of doing this.

See #2576.

vadz committed:
Fix off-by-1 bugs in wxGrid::GetFirstFullyVisible{Row,Column}()

These functions always returned the first partially visible line, not
the first fully visible one as intended because IsXXXShown() always
returned true when it applied to that line and we need to increment the
index first, before calling it -- even if this requires writing the loop
in a slightly uglier way.

This fixes bugs in commits 0920a1646b (Make wxGrid row selecting more
user friendly, 2020-03-04) and 89dd47edee (Make wxGrid column selecting
more user friendly, 2020-03-04).

vadz committed:
Add a comment for ClearSelection()

No real changes, just mention that this function does more than just
clearing the selection.

vadz committed:
Add wxGridSelection::SelectAll() and use it in wxGrid

The difference between calling SelectAll() and SelectBlock() with a
block covering the entire grid is that the former discards any
previously selected blocks, which become clearly redundant.

As a consequence, clicking on the grid corner 10 times in a row still
results in a selection with a single block, not 10 (identical) blocks.

vadz committed:
Stop passing invalid coordinates to ExtendOrCreateCurrentBlock()

This made the logic of this function unnecessarily more complicated.
Instead, just fall back to the current cell coordinates in the only
place where this could happen before.

Doing this still preserves the correct behaviour of Shift-arrow
selection when entire rows/columns are selected and the current cell is
not the leftmost/topmost cell (due to scrolling), but the code is
simpler.

Remove the now always true condition check and assert that it's indeed
always true.

Note that the changes to gridsel.cpp in this commit are best viewed
ignoring whitespace changes.

vadz committed:
Fix Shift-click/arrow behaviour for column/row selection

Extend the current block to the entire line when the corresponding
header is Shift-clicked and, importantly, keep the full-line selection
when using Shift-arrows later to make the selection behave in the
expected way.

vadz committed:
Fix selection expansion in row/column-only selection modes

We should never adjust the fixed coordinate of the selection blocks in
this mode, so ensure we never end up with a partially selected line in
this case by preserving the -- correct by construction -- current block
coordinates in this direction.

vadz committed:
Rename wxGrid::GetSelectionRange() to GetSelectedBlocks()

This seems to be more consistent with the existing functions and doesn't
create ambiguity with a grid range.

Also rename wxGridSelectionRange to just wxGridBlocks as, in principle,
this class could be used for iterating over any blocks, not just the
selected ones.

No changes in functionality, this is just a renaming.

vadz committed:
Also rename wxGridSelection::ExtendOrCreateCurrentBlock()

This function finally doesn't ever create a new block, except for the
trivial case when there is no current block, so rename it to a simpler
and more clear name.

No real changes.

vadz commented:
I pushed my changes and, AFAICS, this should be ready to be merged, i.e. the selection seems to behave as expected from the user point of view, including in the row-only selection mode, and the code is relatively clear.

Please re-test this and let me know about any problems in the selection behaviour. TIA!

vadz commented:
BTW, this is funny: I've realized that the current logic means that when we extend the selection from, say, `R1C1`, to `R1C1:R100C2` by clicking on the second column header, we are going to recognize the row 100 as the "other" row of the selection, so pressing `Shift-Up` arrow will shrink the selection to `R1C1:R99C2`. However if you extend selection from any other cell (except `R100C1`, e.g. `R2C1`, this is not going to happen because the current cell is not on any of the rows of the new block, so pressing `Shift-Up` or `Shift-Down` doesn't do anything at all.

This looked like a bug to me, so I wanted to fix this and make it consistent, however I didn't know what would be better: allow shrinking selection in the second case or not allowing doing it in the first one. So I've decided to check how this works in MS Excel (BTW, if you do this, the following VBA function is pretty handy to follow selection changes:
```vba
Private Sub Workbook_SheetSelectionChange(ByVal Sh As Object, ByVal Target As Range)
  Application.StatusBar = Selection.Address
End Sub
```
) and, to my surprise, it works in _exactly_ the same way as `wxGrid` does with this PR. I.e. there is the same inconsistency there too and, perhaps, for the same reasons.

So, finally, I'm not going to change anything because if this behaviour is good enough for Excel, it's good enough for us too.

OTOH there is still at least one other bug left, with `Shift-Ctrl-arrows` in row/column selection mode, which I will fix soon...

vadz committed:
Fix test for "almost visible" row in GetFirstFullyVisibleRow()

The test didn't correspond to the comment and resulted in always
considering the first row not visible instead of the intended effect.

vadz committed:
Don't use row/column headers for selection incompatible with mode

Mouse events in row/column headers should be just ignored when the
current selection mode disallows row/column selection.

This wasn't the case before, and while actually selecting the line
corresponding to the header was disallowed further down the call chain,
clicking it still unexpectedly cleared the existing selection and
dragging mouse in the header window selected the entire grid.

Fix this by just never entering the corresponding cursor mode in the
first place if it's incompatible with the current selection mode.

vadz committed:
Fix fatal bug in wxGridSelection::DeselectBlock()

The selBlock reference could be used after the vector invalidation,
which resulted in using the wrong vector element at best and could
probably lead to a crash at worst.

Don't use it after invalidating the vector any longer.

This notably fixes a bug in row-or-column selection mode where the grid
wasn't updated visually after modifying its selection by deselecting a
single cell as soon as there were more than one selected row or column.

vadz committed:
Fix Shift-Ctrl-arrows handling

Extending the selection with Ctrl-arrows is different from all the other
cases, as we need to combine both the selection anchor and the current
cell coordinates when doing it.

This means that we can't reuse the same PrepareForSelectionExpansion()
helper for this case, so this function is not useful finally and this
commit removes it entirely. It also replaces GetCurrentBlockCornerRow()
and GetCurrentBlockCornerCol() functions with GetExtensionAnchor() which
combines both of them.

Finally, it adds wxGridDirectionOperations::TryToAdvance() helper to
avoid repeating the IsAtBoundary() check which was previously part of
PrepareForSelectionExpansion() in multiple places.

And because the "extending" and normal parts of DoMoveCursorByBlock()
are so different now, it also factors out AdvanceByBlock() helper which
can be used to keep these parts well separate from each other instead of
intermixing them together.

With all these preparatory changes, it's finally possible to implement
the "extending selection by block" logic relatively easily, with the
bulk of this branch actually taken by comments explaining why do we have
to do what we do.

Add unit tests verifying that the functions used by Shift-Ctrl-arrow
work as expected.

vadz committed:
Allow extending selection using Shift-Page Up/Down keys

Also make Page Up/Down themselves work consistently with the other
cursor movement keys and clear current selection if they move the
cursor.

Even though DoMoveCursorByPage() is simpler than DoMoveCursorByBlock(),
still factor out AdvanceByPage() for consistency with AdvanceByBlock()
and because it still makes the code more clear.

vadz committed:
Disallow shrinking rows/columns in row-or-column selection mode

Fix the check for disallowing changing block rows/columns to take into
account the row-or-column selection mode too.

vadz committed:
Refresh the grid correctly when the selection mode changes

Update the cells that had to be deselected due to the selection mode
change.

This used to work, but the refresh was lost during a previous
refactoring.

vadz committed:
Avoid horizontal scrolling when pressing Shift-Page Up/Down

Scroll the new end of selection into view only vertically, it would be
unexpected for the vertical Page Up/Down movement to scroll the grid
horizontally.

vadz committed:
Make the line whose header was Ctrl-clicked current

This is more consistent with Ctrl-clicking cells.

It also slightly improves behaviour when Ctrl-dragging mouse over the
range of selected lines, although it's still somewhat surprising.

vadz committed:
Do make Ctrl-clicked cell current

There isn't really any valid reason to ever set the current cell to the
corner of a block remaining after breaking up the currently selected
block after the deselection of the just clicked cell, so don't do it.

Just set the current cell to the clicked cell, whether it's selected or
not.

vadz committed:
Stop setting current cell on first Ctrl-drag event

This is not necessary any longer as the current cell is changed on
Ctrl-click since the previous commit.

This commit is best viewed ignoring whitespace.

vadz committed:
Simplify and improve wxGrid::DoGridCellDrag() return value logic

Remove the "performDefault" variable which didn't really seem to help,
as the actual meaning of the return value is whether we should start
drag-selecting or not and the name of this variable didn't reflect this
anyhow.

Return false if we are over an invalid cell: this shouldn't change
anything when we're already dragging, but would prevent us from starting
to drag from an invalid cell which seems more correct, even if it's not
clear if this can happen in practice.

Also move hiding the edit control inside "isFirstDrag" check, there is
no reason to do it on every drag.

vadz committed:
Don't allow drag-extending selection from deselected cell

Dragging in a grid with Ctrl key pressed starting from a previously
selected cell behaved very counterintuitively if not downright wrongly,
as the selection logic assumes that the selection anchor itself is
always selected, which wasn't true in this case.

Solve the problem by just not extending the selection at all when
starting to drag from a deselected cell. This means that Ctrl-dragging
doesn't do anything any more, but it's not a huge loss and to make it
work well while still allowing to use Ctrl-click to toggle the cell
selection, we'd need to implement a whole new and different
drag-deselect mode, as is done in Microsoft Excel 2016 (note that the
previous versions of Excel don't implement Ctrl-dragging neither).

vadz committed:
Implement alternative solution to Ctrl-drag problem

This commit doesn't change the behaviour compared to the previous one,
but provides an alternative implementation of the same goal, which seems
preferable: instead of not extending the selection while Ctrl-dragging,
just don't enter dragging mode, i.e. don't capture the mouse and don't
set m_isDragging to true, if we start it from a previously selected, and
hence currently deselected, cell.

vadz committed:
Disable extending selection with Ctrl-drag for rows/columns too

Don't extend the selection if the anchor line is not selected, as this
doesn't work correctly because the entire selection logic supposes the
anchor itself is selected and, moreover, it's not really clear how could
it would otherwise.

This commit does the same thing for rows/columns as the grandparent
commit did to the cells. Unfortunately a somewhat cleaner solution of
the parent commit can't be easily applied to the existing rows/columns
code and it's arguably not worth changing it in depth just for this.

vadz committed:
Optimize test for the row/column selection

Use wxGridSelection::IsInSelection() instead of GetSelectedRows/Cols()
which can be much slower as they need to produce an array containing
indices of all the selected rows/columns.

vadz committed:
Make Ctrl/Shift-Space apply to all cells of the current selection

This provides a convenient way to select multiple lines and is
consistent with the operation of these keys in spreadsheet programs.

thesiv commented on interface/wx/grid.h:
```diff
@@ -1816,6 +1816,250 @@ class wxGridCellCoords
+        @return
+            @true if @a other is entirely contained within this block.
+     */
+    int ContainsBlock(const wxGridBlockCoords& other) const;
```
```suggestion
    bool ContainsBlock(const wxGridBlockCoords& other) const;
```

vadz replied:
Fixed, thanks.

vadz committed:
Fix documented return type of wxGridBlockCoords::ContainsBlock()

This should have been part of 791a9e68ae (Rename and simplify
wxGridBlockCoords::ContainsBlock(), 2020-04-05).

vadz committed:
Make wxGridSelection::IsInSelection() const

There is really no reason for it not to be.

vadz committed:
Fix extending selection starting from unselected current cell

Unselected current cell should always be considered as the current
selection block to extend, as it doesn't make sense to extend any other
block (perhaps selected on another side of the grid) when pressing
Shift-arrow.

This scenario could be achieved by selecting a block and Ctrl-clicking a
cell (either inside or outside the selection) twice and then extending
it using Shift-arrow keys. Previously, this behaved in a strange way,
combining the corner of the selected block with the target of the
movement, whereas now this just starts selecting a new block from the
current cell as expected.

vadz committed:
Always explicitly set split orientation in DeselectBlock()

No real changes, just make the code more clear by always explicitly
selecting either wxHORIZONTAL or wxVERTICAL instead of default for the
former for no good reason (by symmetry, it is not a better choice than
wxVERTICAL).

Also list all wxGridSelection enum elements in the switch over selection
mode to avoid warnings about not handling wxGridSelectRows.

vadz committed:
Allow selecting rows/columns in row-or-column selection mode

Don't blankly forbid selecting any blocks at all in this mode, this
didn't really make any sense.

Moreover, SelectBlock() not doing anything prevented wxGrid code
handling {Ctrl,Shift}-Space from doing anything in this mode and, worse,
broke the logic of DeselectBlock() which relied on using SelectBlock()
to select the remaining parts of the selection, so this commit fixes
using Ctrl-click for deselecting rows or columns in this selection mode,
which was previously completely broken.
